File: js/cozy_spoke.js
// https://github.com/cozy-comfyui/cozy_spoke

import { api } from "../../../scripts/api.js";
import { app } from "../../../scripts/app.js";

const _id = "Action"
const _root = "cozy_spoke"
const EVENT_COZY_UPDATE = "cozy-event-combo-update";

const nodeWidgetHandlers = [
    {
        "parent_widget": "act_type",
        "child_widget": "act",
    }
]

async function apiRoute(route, data = null, id = null) {
    const full_route = `/${_root}/${route}`;
    var blob = {
        method: "GET",
        headers: { "Content-Type": "application/json" }
    }

    // if we are passing data, or need a specific node by ID, we must use a POST
    if (data != null || id != null) {
        blob['method'] = "POST",
            blob['body'] = JSON.stringify({
                id: id,
                data: data
            });
    }

    try {
        const response = await api.fetchApi(`${full_route}`, blob);
        if (!response.ok) {
            throw new Error(`Error: ${response.status} - ${response.statusText}`);
        }
        return response;

    } catch (error) {
        console.error("API call to Jovimetrix failed:", error);
        throw error; // or return { success: false, message: error.message }
    }
}

app.registerExtension({
    name: 'cozy.spoke.' + _id,
    async beforeRegisterNodeDef(nodeType, nodeData) {
        if (nodeData.name !== _id) {
            return
        }

        const onNodeCreated = nodeType.prototype.onNodeCreated;
        nodeType.prototype.onNodeCreated = async function () {
            const me = onNodeCreated?.apply(this);

            // Update all children widgets according to the parent widgets value.
            // in respect of what's defined in the config files
            nodeWidgetHandlers.forEach((handler) => {
                const parentWidget = this.widgets.find(w => w.name == handler.parent_widget);
                const childWidget = this.widgets.find(w => w.name == handler.child_widget);

                parentWidget.callback = async () => {
                    try {
                        const response = await apiRoute("node", parentWidget.value);
                        const responseData = await response.json();
                        childWidget.options.values = responseData.data;
                        childWidget.value = responseData.data[0] || "random";
                        this.setDirtyCanvas(true, true);
                    } catch (error) {
                        console.error('Error in parent widget callback:', error);
                    }
                }
            });

            // process the message sent from python.
            // this could be during node execution or outside of it.
            async function python_update(data) {
                // Assuming you will do something with the data from Python here
                console.info('Received update from Python for child widget:', data);
            }

            // catch the event(s) from python.
            api.addEventListener(EVENT_COZY_UPDATE, python_update);

            // clean up your mess when leaving.
            this.onDestroy = () => {
                api.removeEventListener(EVENT_COZY_UPDATE, python_update);
            };
            return me;
        }
    }
})

File: py/lib/cozy_spoke.py
# CozySpoke – Widget manipulation with API
# https://github.com/cozy-comfyui/cozy_spoke

import time
# import toml
# from pathlib import Path
from typing import Any
from aiohttp import web  # , ClientSession
from server import PromptServer

from ..node import Node

EVENT_COZY_UPDATE = "cozy-event-combo-update"


class TimedOutException(Exception):
    pass


class ComfyAPIMessage:
    """
    This is to collect messages from JS for nodes that are parsing/looking
    for messages.
    """
    MESSAGE = {}

    @classmethod
    def poll(cls, ident, period=0.01, timeout=3) -> Any:
        _t = time.monotonic()
        if isinstance(ident, (set, list, tuple, )):
            ident = ident[0]

        sid = str(ident)
        while not (sid in cls.MESSAGE) and time.monotonic() - _t < timeout:
            time.sleep(period)

        if not (sid in cls.MESSAGE):
            raise TimedOutException
        dat = cls.MESSAGE.pop(sid)
        return dat


def initialize_cozy_spoke():

    data = Node.load_data("actions.toml")

    @PromptServer.instance.routes.get("/cozy_spoke")
    async def route_cozy_spoke(request) -> Any:
        """A simple lookup for the data provided.
        The function names need to be unique for each 'route'.
        Returns all messages stored in the Message Bus.
        """
        return web.json_response(ComfyAPIMessage.MESSAGE)

    @PromptServer.instance.routes.post("/cozy_spoke")
    async def route_cozy_spoke_combo(request) -> Any:
        """A catch all route to pass messages for specific node messages needed
        during node execution. The node itself will search the message bucket
        and can process during its run function call."""
        json_data = await request.json()
        if (did := json_data.get("id")) is not None:
            # stores the data-call for the instanced node (by id)
            # to check on execute
            ComfyAPIMessage.MESSAGE[str(did)] = json_data
            return web.json_response(json_data)
        return web.json_response({})

    @PromptServer.instance.routes.post("/cozy_spoke/node")
    async def route_cozy_spoke_update(request) -> Any:
        """A specific route for a specific node
        to process outside of an execution run.
        Here we manipulate the values output based on the selection in ComboA
        """

        # Show/Hide widgets according to data dict
        json_data = await request.json()
        result = {}
        if (response := json_data.get("data")) is not None:
            acts = data["nsfw"]["acts"]
            selected_acts = acts.get(response.lower(), [])
            inputs = Node.build_inputs_list(selected_acts)
            result = {'data': inputs}

        return web.json_response(result)
